src/agent/storage-queue/src/local_queue.rs

```diff
@@ -0,0 +1,147 @@
+        let send_data = || async {
+            let mut buffer = Vec::new();
+            serde_xml_rs::to_writer(&mut buffer, &data)
+                .map_err(|_| anyhow::anyhow!("unable to deserialize"))?;
+            let mut locked_q = self
+                .queue
+                .lock()
+                .map_err(|_| anyhow::anyhow!("unable to acquire lock"))?;
+            locked_q
+                .add(buffer.as_slice())
+                .map_err(|_| anyhow::anyhow!("unable to queue message"))?;
+            Ok(())
+        };
+
+        let backoff = ExponentialBackoff {
+            current_interval: SEND_RETRY_DELAY,
+            initial_interval: SEND_RETRY_DELAY,
+            max_elapsed_time: Some(MAX_ELAPSED_TIME),
+            ..ExponentialBackoff::default()
+        };
+        let notify = |err, _| println!("IO error: {}", err);
+        retry_notify(backoff, send_data, notify).await?;
+
+        Ok(())
+    }
+
+    pub async fn pop(&self) -> Result<Option<LocalQueueMessage>> {
+        let receive_data = || async {
+            let mut locked_q = self
+                .queue
+                .lock()
+                .map_err(|_| anyhow::anyhow!("unable to acquire lock"))?;
+            let data = locked_q
+                .peek()
+                .map_err(|_| anyhow::anyhow!("unable to peek"))?;
+            locked_q
+                .remove()
+                .map_err(|_| anyhow::anyhow!("unable to pop message"))?;
+
+            let message = data.map(|d| LocalQueueMessage { data: d.to_vec() });
+            Ok(message)
+        };
+
+        let backoff = ExponentialBackoff {
+            current_interval: SEND_RETRY_DELAY,
+            initial_interval: SEND_RETRY_DELAY,
+            max_elapsed_time: Some(MAX_ELAPSED_TIME),
+            ..ExponentialBackoff::default()
+        };
+        let notify = |err, _| println!("IO error: {}", err);
+        let result = retry_notify(backoff, receive_data, notify).await?;
+
+        Ok(result)
+    }
+}
+
+use tokio::sync::mpsc::{
+    error::TryRecvError, unbounded_channel, UnboundedReceiver, UnboundedSender,
+};
+
+/// Queue based on mpsc channel
+#[derive(Debug, Clone)]
+pub struct ChannelQueueClient {
+    sender: Arc<Mutex<UnboundedSender<Vec<u8>>>>,
+    receiver: Arc<Mutex<UnboundedReceiver<Vec<u8>>>>,
+    pub url: reqwest::Url,
+}
+
+impl ChannelQueueClient {
+    pub fn new() -> Result<Self> {
+        let (sender, receiver) = unbounded_channel();
+        Ok(ChannelQueueClient {
+            sender: Arc::new(Mutex::new(sender)),
+            receiver: Arc::new(Mutex::new(receiver)),
+            url: reqwest::Url::parse("mpsc://channel")?,
+        })
+    }
+
+    pub async fn enqueue(&self, data: impl Serialize) -> Result<()> {
+        let sender = self
+            .sender
+            .lock()
+            .map_err(|_| anyhow::anyhow!("unable to acquire lock"))?;
+        let mut buffer = Vec::new();
+        serde_xml_rs::to_writer(&mut buffer, &data)
```

Comment:
Can you speak to why you chose XML as the local queue serialization?

Reply:
I chose xml to be compatible with azure. The body of message in the queue is [in xml](https://docs.microsoft.com/en-us/rest/api/storageservices/get-messages#response-body)

Reply:
never mind I thought i was looking at azure queue.
This was probably the result of a copy paste from the azure queue. I can convert to json instead

Reply:
i reverted this back to xml because of the serialisation behavior of simple string. 
the json serializer puts quotes around the string which breaks the file monitoring currently